google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/ChangeStreamMutation.java

```diff
@@ -0,0 +1,293 @@
+ *
+ * <p>Case 1) User initiated mutation.
+ *
+ * <pre>{@code
+ * ChangeStreamMutation.Builder builder = ChangeStreamMutation.createUserMutation(...);
+ * builder.setCell(...);
+ * builder.deleteFamily(...);
+ * builder.deleteCells(...);
+ * ChangeStreamMutation changeStreamMutation = builder.setToken(...).setLowWatermark().build();
+ * }</pre>
+ *
+ * Case 2) Garbage Collection mutation.
+ *
+ * <pre>{@code
+ * ChangeStreamMutation.Builder builder = ChangeStreamMutation.createGcMutation(...);
+ * builder.setCell(...);
+ * builder.deleteFamily(...);
+ * builder.deleteCells(...);
+ * ChangeStreamMutation changeStreamMutation = builder.setToken(...).setLowWatermark().build();
+ * }</pre>
+ *
+ * Make this class non-final so that we can create a subclass to mock it.
+ */
+@InternalExtensionOnly("Intended for use by the BigtableIO in apache/beam only.")
+@AutoValue
+public abstract class ChangeStreamMutation implements ChangeStreamRecord, Serializable {
+  private static final long serialVersionUID = 8419520253162024218L;
+
+  public enum MutationType {
+    USER,
+    GARBAGE_COLLECTION
+  }
+
+  private static ChangeStreamMutation create(Builder builder) {
+    return new AutoValue_ChangeStreamMutation(
+        builder.rowKey,
+        builder.type,
+        builder.sourceClusterId,
+        builder.commitTimestamp,
+        builder.tieBreaker,
+        builder.token,
+        builder.lowWatermark,
+        builder.entries.build());
+  }
+
+  /**
+   * Creates a new instance of a user initiated mutation. It returns a builder instead of a
+   * ChangeStreamMutation because `token` and `loWatermark` must be set later when we finish
+   * building the logical mutation.
+   */
+  static Builder createUserMutation(
+      @Nonnull ByteString rowKey,
+      @Nonnull String sourceClusterId,
+      @Nonnull Timestamp commitTimestamp,
+      int tieBreaker) {
+    return new Builder(rowKey, MutationType.USER, sourceClusterId, commitTimestamp, tieBreaker);
+  }
+
+  /**
+   * Creates a new instance of a GC mutation. It returns a builder instead of a ChangeStreamMutation
+   * because `token` and `loWatermark` must be set later when we finish building the logical
+   * mutation.
+   */
+  static Builder createGcMutation(
+      @Nonnull ByteString rowKey, @Nonnull Timestamp commitTimestamp, int tieBreaker) {
+    return new Builder(rowKey, MutationType.GARBAGE_COLLECTION, "", commitTimestamp, tieBreaker);
+  }
+
+  //  private void readObject(ObjectInputStream input) throws IOException, ClassNotFoundException {
+  //    input.defaultReadObject();
+  //
+  //    @SuppressWarnings("unchecked")
+  //    ImmutableList<Entry> deserialized = (ImmutableList<Entry>) input.readObject();
+  //    getEntries() = ImmutableList.<Entry>builder().addAll(deserialized);
+  //  }
+  //
+  //  private void writeObject(ObjectOutputStream output) throws IOException {
+  //    output.defaultWriteObject();
+  //    output.writeObject(getEntries());
+  //  }
```

Comment:
whats going on here?

Reply:
I lost my brain power to clean this up after some refactoring.. 

Done.